cave/plot/algorithm_footprint.py

```diff
@@ -215,36 +278,177 @@ def get_2NN(x, X):
+            output path
         """
-        start = time.time()
-        self.algo_labels = {a:{} for a in self.algorithms}
-        for i in self.insts:
-            performances = [self.get_performance(a, i) for a in self.algorithms]
-            best_performance = min(performances)
-            for a in self.algorithms:
-                performance = self.get_performance(a, i)
-                #self.logger.debug("%s on \'%s\': best/this (%f/%f=%f)",
-                #                  self.algo_names[a], i,
-                #                  best_performance, performance,
-                #                  best_performance / performance)
-                if (performance == 0 or
-                    (best_performance/performance >= epsilon and
-                     not performance >= self.cutoff)):
-                    # Algorithm for instance is in threshhold epsilon
-                    #   and no timeout
-                    label = 1
-                else:
-                    label = 0
-                self.algo_labels[a][i] = label
-        self.logger.debug("Labeling instances in %.2f secs.", time.time() - start)
+        if len(insts) == 0:
+            insts = self.insts
+
+        good_idx, bad_idx = [], []
+        for k, v in self.algo_performance[conf].items():
+            # Only consider passed insts
+            if not k in insts:
+                continue
+            # Append inst-idx either to good or to bad
+            if self.algo_labels[conf][k] == 0:
+                bad_idx.append(self.insts.index(k))
+            else:
+                good_idx.append(self.insts.index(k))
+        assert(len(good_idx) == len(set(good_idx)))
+        assert(len(bad_idx) == len(set(bad_idx)))
+        good_idx, bad_idx = np.array(good_idx), np.array(bad_idx)
+        self.logger.debug("for config %s good: %d, bad: %d",
+                          self.algo_names[conf], len(good_idx), len(bad_idx))
+        return (good_idx, bad_idx)
+
+    def plot2d(self):
+        """ Plot shaded 2d-version of the algorithm footprint. """
+        plots = []
+        for a in self.algorithms:
+            # Plot without clustering (for all insts)
+            out_fn = os.path.join(self.output_dir, 'footprint_' +
+                                  self.algo_names[a] + '_2d.png')
+            self.logger.debug("Plot saved to '%s'", out_fn)
+            fig, ax = plt.subplots()
+            good_idx, bad_idx = self._get_good_bad(a)
+            # As we don't have such a high resolution when plotting, i.e. we don't see differences between 0.001 and 0.00001
+            # all points that lie close by might overlap completely. To easily spot these, squash everything down to one
+            # decimal
+            good = np.array([self.features_2d[idx] for idx in good_idx])
+            bad = np.array([self.features_2d[idx] for idx in bad_idx])
+            good, bad = np.around(np.array(good), decimals=1), np.around(np.array(bad), decimals=1)
```

Comment:
I'm still not too sure if always rounding to 1 decimal point is the correct way. PCA could show instances to be in a range of 0.15 to 0.39. If we would round every point we might group points together, that would not overlap in this small range. It really depends on the range of the resulting PCA.

Reply:
I created a new issue, so I can merge this PR.

Reply:
:+1: